File: Documents/GitHub/Web-Projects/project5/calc_statistics.js
// finds mean
export function mean(data) {
    let sum = (a,b) => a + b;
    let len = data.length;

    if (len == 0) {
        return null;
    }
    
    let sumdata = data.reduce(sum);
    
    return (sumdata / len);
}

// comparator function
export function compareNumbers(a,b) {
    return a - b;
}

// find median
export function median(data) {
    let len = data.length;
    if (len == 0) {
        return null;
    }
    let datasorted = data.sort(compareNumbers);
    if (len % 2 == 0) {
        return (datasorted[len/2 - 1] + datasorted[len/2]) / 2;
    } else {
        return (datasorted[len/2 - 1]);
    }
}

// finds max in data set
export function maximum(data) {
    let max = data[0]

    for (let i = 1; i < data.length; i++) {
        if (data[i] > max) {
            max = data[i];
        }
    }
    return max;
}

// finds smallest in data set
export function find_smallest(data) {
    let min = data[0]

    for (let i = 1; i < data.length; i++) {
        if (data[i] < min) {
            min = data[i];
        }
    }
    return min;
}

// finds mode
export function mode(data) {
    const frequencyMap = {};
    data.forEach(element => {
        frequencyMap[element] = (frequencyMap[element] || 0) + 1;
    });

    let modes = [];
    let maxFreq = 0;

    for (const num in frequencyMap) {
        let freq = frequencyMap[num];
        if (freq > maxFreq) {
            maxFreq = freq;
            modes = [parseFloat(num)];
        } else if (freq === maxFreq) {
            modes.push(parseFloat(num));
        }
    }

    return modes;
}

const data = [4,6];
console.log(mean(data));
console.log(mean([4,6]));

File: Documents/GitHub/Web-Projects/project2/calc_statistics.rb
# Various stats functions
module CalcStatistics
  # Calculates the mean of an array of numbers
  # @param data [Array<Numeric>] An array of numbers to calculate the mean of
  # @return [Float] The mean of the array inputted
  def mean(data)
    return nil if data.empty? # Return nil if the array is empty
  
    sum = data.reduce(0, :+)
    mean = sum.to_f / data.size
    mean
  end
  

  # Finds the maximum value of an array of numbers
  # @param data [Array<Numeric>] An array of numbers to find the maximum of
  # @return [Float] The maximum value of the array inputted
  def maximum(data) 
    # initialize max to first value of array
    max = data[0]

    # set value in array to max if greater than current max
    data.each{ |val|
      if val > max
        max = val
      end
    }
    # return max
    max
  end

  # Finds the median value of an array of numbers
  # @param data [Array<Numeric>] An array of numbers to find the median of
  # @return [Float, Numeric] The median value of the array inputted
  def median(data)
    sorted = data.sort
    puts "#{sorted}"
    length = sorted.length
    if length % 2 == 1
      sorted[length / 2] # odd length, return middle element
    else
      mid1 = sorted[length / 2 - 1]
      mid2 = sorted[length / 2]
      (mid1 + mid2) / 2.0 # even length, return average of middle elements
    end
  end

    # Function that returns the smallest value from a set of numbers
  # @param numbers [Array<Numeric>] The array of numbers to be checked
  # @return [Numeric] The smallest number in the array
  def find_smallest(numbers)
    smallest = numbers[0] # assume the first number is the smallest

    numbers.each do |num|
      if num < smallest
        smallest = num
      end
    end
    return smallest
  end

    # Function that finds the mode of a dataset
  # @param data [Array<Numeric>] An array of numbers to find the mode of
  # @return [Float, Numeric] The mode value of the array inputted
  def mode(data)
    frequency = Hash.new(0)
    max_frequency = 0
    mode = nil

    data.each do |num|
      frequency[num] += 1
      if frequency[num] > max_frequency
        max_frequency = frequency[num]
        mode = num
      end
    end

    mode
  end

end